src/changelog/prepareChangelogPaths.ts

```diff
@@ -0,0 +1,142 @@
+import fs from 'fs';
+import path from 'path';
+import { v4 as uuid } from 'uuid';
+import type { BeachballOptions } from '../types/BeachballOptions';
+
+interface ChangelogPaths {
+  /** Absolute path to changelog md file (new or existing) */
+  md?: string;
+  /** Absolute path to changelog json file (new or existing) */
+  json?: string;
+  /** Suffix being used in files, if any */
+  suffix?: string;
+}
+
+/**
+ * Get the paths to the changelog files. Also handles conversion between `changelog.uniqueFilenames`
+ * being true and false/unset (moving the files if needed).
+ *
+ * @returns object with each changelog path, or undefined if that changelog shouldn't be written.
+ */
+export function prepareChangelogPaths(params: {
+  options: Pick<BeachballOptions, 'changelog' | 'generateChangelog'>;
+  /** cwd where changelogs are located */
+  changelogAbsDir: string;
+}): ChangelogPaths {
+  const { options, changelogAbsDir: cwd } = params;
+  const { changelog, generateChangelog } = options;
+  const { uniqueFilenames } = changelog || {};
+
+  const paths: ChangelogPaths = {};
+  let existingSuffixedPaths: ChangelogPaths | undefined;
+
+  if (!generateChangelog) {
+    return paths;
+  }
+
+  for (const ext of ['md', 'json'] as const) {
+    if (generateChangelog !== true && generateChangelog !== ext) {
+      continue; // not generating this file type
+    }
+
+    const defaultPath = path.join(cwd, `CHANGELOG.${ext}`);
+
+    if (uniqueFilenames) {
+      // Filenames should use a suffix. Start by getting any existing suffixed changelogs...
+      existingSuffixedPaths ??= _getExistingSuffixedChangelogs(cwd);
+      // Use the existing suffix or generate a new one
+      paths.suffix ??= existingSuffixedPaths.suffix ?? uuid().slice(0, 8);
```

Comment:
This use of a "random" suffix is perhaps concerning to me. The method prototype doesn't currently offer anything else, but I was thinking that having a deterministic suffix based on the directory for the CHANGELOG files (relative to src root) or the package name would be helpful.

The reason to have something like that is to avoid possible collisions if two changelogs are generated in parallel somehow and no longer conflict on merge. Or if there is somehow a bug in the discovery of existing suffixed files, the deterministic algorithm would produce the same filename as the previous time.

It's worth being a bit of a stickler here, since the downside is quite real: if these suffixes start being added randomly, then we are worse off than before in terms of Git finding good deltas. I'd rate this as "low probability to be wrong, but high impact if wrong".

Reply:
Thanks, I was sort of wondering about the random suffix vs. hashing something. I think if we're actually hashing something, it should be the package name, since that's probably more stable than the path. I'll just need to find a different library for hashing (`uuid` is only for random GUIDs).